### ISSUE TITLE
Weird 3D Point Cloud

### ISSUE THREAD

**Issue by Jayden9912** [nianticlabs/monodepth2#366]
Hi. I have been looking for mistakes for hours. Any help would be much appreciated.
I am trying to plot 3D point cloud by using the points get from backproject_depth.
This is the K I am using:
`#original image size 1242x375
scale = 0
w,h = 640,192

K = np.array([[0.58, 0, 0.5, 0],
                           [0, 1.92, 0.5, 0],
                           [0, 0, 1, 0],
                           [0, 0, 0, 1]], dtype=np.float32)

K[0, :] *= w // (2 ** scale) #Kitti
K[1, :] *= h // (2 ** scale) #Kitti
inv_K = np.linalg.pinv(K)
inv_K = np.expand_dims(inv_K,0)
inv_K = torch.from_numpy(inv_K).to(device)

backproject_depth = BackprojectDepth(1, h, w)
backproject_depth.to(device)`

The code below is modified from test_simple:
`with torch.no_grad():
    for idx, image_path in enumerate(paths):

        if image_path.endswith("_disp.jpg"):
            # don't try to predict disparity for a disparity image!
            continue

        # Load image and preprocess
        input_image = pil.open(image_path).convert('RGB')
        original_width, original_height = input_image.size
        input_image = input_image.resize((feed_width, feed_height), pil.LANCZOS)
        input_image = transforms.ToTensor()(input_image).unsqueeze(0)
        
        # PREDICTION
        input_image = input_image.to(device)
        features = encoder(input_image)
        outputs = depth_decoder(features)

        disp = outputs[("disp", 0)]
        disp_resized = torch.nn.functional.interpolate(
        disp, (original_height, original_width), mode="bilinear", align_corners=False)
        
        scaled_disp, depth = disp_to_depth(disp, 0.1, 100)
        
        cam_points = backproject_depth(depth, inv_K)
        cam_points = cam_points.cpu().numpy()[0].T[:, :3].astype(np.float32)
        cam_points[:,2]*30
        
        pcd = o3d.geometry.PointCloud()
        pcd.points = o3d.utility.Vector3dVector(cam_points)
        o3d.visualization.draw_geometries([pcd]) # Visualize the point cloud   
        
        disp_resized_np = disp_resized.squeeze().cpu().numpy()
        vmax = np.percentile(disp_resized_np, 95)
        normalizer = mpl.colors.Normalize(vmin=disp_resized_np.min(), vmax=vmax)
        mapper = cm.ScalarMappable(norm=normalizer, cmap='magma')
        colormapped_im = (mapper.to_rgba(disp_resized_np)[:, :, :3] * 255).astype(np.uint8)
        plt.imshow(colormapped_im)`

And this is the point cloud I got:
![image](https://user-images.githubusercontent.com/85933053/122030349-494c6f80-ce00-11eb-9d8c-850cc6a607c2.png)
![image](https://user-images.githubusercontent.com/85933053/122030446-5bc6a900-ce00-11eb-9841-80afa5d36f9d.png)

This is the disparity map I got:
![image](https://user-images.githubusercontent.com/85933053/122030541-739e2d00-ce00-11eb-80db-af33c130ceaa.png)

The kitti data I used is kitti_data/2011_09_26/2011_09_26_drive_0001_sync/image_00/data/0000000000.jpg
The model trained is mono_model with size 640x192.

Is there any mistake in my code or concept. Any discussion is welcomed. Thanks in advance.

**Comment by mdfirman** [comment]
Hi @Jayden9912  – thanks for trying this out and sharing your results.

I have done this process before, of 3D visualisation. And I'm actually not sure there are any mistakes in what you are doing.

Looking at the 3D point cloud, it looks to be in reasonable alignment with the depth map you've shared.
Here I've drawn some pink lines matching up regions in the point cloud with those in the depth image

![image](https://user-images.githubusercontent.com/6114258/123061907-f7a97380-d403-11eb-8c28-3034d5ab60f7.png)

As I note above: Sky is usually pretty badly predicted in these sort of predictions, as the self-supervised losses don't work too well. So it's being predicted here with too small a depth, causing it to come 'forward' in the point cloud.

Here are some visualisation techniques which might help to produce more understandable point clouds:

* Colour each point in the point cloud with the colour of the RGB pixel. 
* Colour each point in the point cloud with colour indicating the semantic class of the pixel (run a semantic segmentor first to get the segmentation map)
* Remove sky pixels before creating point cloud
* Remove sky and ground pixels before creating point cloud, leaving only objects like trees and cars
* Try this but for an image with more foreground objects, e.g. cars and bicycles close to the camera.


Let us know how you get on! Thanks

**Comment by Jayden9912** [comment]
@mdfirman Cool. Thanks for your detailed explanation. I thought I did wrongly. I will explore more on the suggestion given. Closing this for now.
